Better docs for standard topology labels (#24783)
kubernetes/website: content/en/docs/reference/kubernetes-api/labels-annotations-taints.md

if you want to target certain workloads to certain instance types, but typically you want
to rely on the Kubernetes scheduler to perform resource-based scheduling. You should aim to schedule based on properties rather than on instance types (for example: require a GPU, instead of requiring a `g2.2xlarge`).

## topology.kubernetes.io/region {#topologykubernetesioregion}

Example:

`topology.kubernetes.io/region=us-east-1`

See [topology.kubernetes.io/zone](#topologykubernetesiozone).

## topology.kubernetes.io/zone {#topologykubernetesiozone}

Example:

`topology.kubernetes.io/zone=us-east-1c`

Used on: Node, PersistentVolume

> On Node: The `kubelet` populates this with the information as provided by the `cloudprovider`.  This will be set only if you are using a `cloudprovider`. However, you should consider setting this on nodes if it makes sense in your topology.

Reviewer: This should probably mention cloud-controller-manager for the external cloud provider case:
```
On Node: The `kubelet` or the external `cloud-controller-manager` populates... 
```

Author: ACK

> On PersistentVolume: The `PersistentVolumeLabel` admission controller automatically adds zone labels to PersistentVolumes, on GCE and AWS.

Reviewer: I don't think this is true, we actually opted-out of supporting the GA labels via the PVL admission controller. More context in this thread https://github.com/kubernetes/kubernetes/pull/81431#issuecomment-525435518. This may have changed last I checked though. cc @msau42

Reviewer: Ah I recall the reasoning now -- it's because volume providers were updated to set `PV.Spec.NodeAffinity` which is the preferred API over labels.

Reviewer: So I think the update required for this PR would be something like:
```
On PersistentVolume: topology-aware volume provisioners will automatically set node affinity constraints to PersistentVolumes
```

Author: Can you propose better wording?   Or just strike this entirely?

Reviewer 2: The pv label admission controller also sets nodeaffinity but with beta labels too. At the time, we thought this would be fine and we can use csi migration to properly translate the beta labels during scheduling. However I agree this could be confusing and alarming for users that see their PVs provisioned with deprecated labels, and with csi migration timelines slipping, we should probably revisit that decision.

I added some thoughts here: https://github.com/kubernetes/kubernetes/issues/92237#issuecomment-719709773

Reviewer: > Do I interpret correctly that the old labels will be respected but the new will not?

Only if the labels are used -- new labels should still be respected in `PV.Spec.NodeAffinity` which is the desired API for volume topology as opposed to labels. So I think just wording the docs based on my comment above https://github.com/kubernetes/website/pull/24783#discussion_r514841131 would be sufficient.

Author: I'll adopt Andrew's words.  PTAL

> On PersistentVolume: The `PersistentVolumeLabel` admission controller will automatically add zone labels to PersistentVolumes, on GCE and AWS.

Reviewer 3: nit:
```suggestion
On PersistentVolume: The `PersistentVolumeLabel` admission controller automatically adds zone labels to PersistentVolumes, on GCE and AWS.
```

Author: ACK

> A zone represents a logical failure domain.  It is common for kubernetes clusters to span multiple zones for increased availability.  While the exact definition of a zone is left to infrastructure implementations, common properties of a zone include very low network latency within a zone, no-cost network traffic within a zone, and failure independence from other zones.  For example, nodes within a zone might share a network switch, but nodes in different zones should not.

Reviewer 3: nit:
```suggestion
A zone represents a logical failure domain.  It is common for Kubernetes clusters to span multiple zones for increased availability.  While the exact definition of a zone is left to infrastructure implementations, common properties of a zone include very low network latency within a zone, no-cost network traffic within a zone, and failure independence from other zones.  For example, nodes within a zone might share a network switch, but nodes in different zones should not.
```
(aside: this could become a glossary definition and then be reused via a shortcode).

Author: ACK the caps.

example how to do that?

> A region represents a larger domain, made up of one or more zones.  It is uncommon for kubernetes clusters to span multiple reqions,  While the exact definition of a zone or region is left to infrastructure implementations, common properties of a region include higher network latency between them than within them, non-zero cost for network traffic between them, and failure independence from other zones or regions.  For example, nodes within a region might share power infrastructure, but nodes in different regions probably should not.

Reviewer 3: How about:
```suggestion
A region represents a larger domain, made up of one or more zones.  It is uncommon for Kubernetes clusters to span multiple regions,  While the exact definition of a zone or region is left to infrastructure implementations, common properties of a region include higher network latency between them than within them, non-zero cost for network traffic between them, and failure independence from other zones or regions. For example, nodes within a region might depend on the same electrical grid, but nodes in different regions typically would not.
```

Author: "power infra" is meant to refer to one of grid, substation, generator, UPS, etc.

All other changed ACK.

> A region represents a larger domain, made up of one or more zones.  It is uncommon for Kubernetes clusters to span multiple regions,  While the exact definition of a zone or region is left to infrastructure implementations, common properties of a region include higher network latency between them than within them, non-zero cost for network traffic between them, and failure independence from other zones or regions.  For example, nodes within a region might share power infrastructure (e.g. a UPS or generator), but nodes in different regions typically would not.

Reviewer 4: Also, it might be worth calling out that a region need not be a physical/geographical location?

> A region represents a larger domain, made up of one or more zones.  It is uncommon for Kubernetes clusters to span multiple regions,  While the exact definition of a zone or region is left to infrastructure implementations, common properties of a region include higher network latency between them than within them, non-zero cost for network traffic between them, and failure independence from other zones or regions.  For example, nodes within a region might share power infrastructure (e.g. a UPS or generator), but nodes in different regions typically would not.

Reviewer 3: If a region isn't linked to a physical/geographical location, is it right to call it a “region”?

Kubernetes makes a few assumptions about the structure of zones and regions:
1) regions and zones are hierarchical: zones are strict subsets of regions and no zone can be in 2 regions

> 2) zone names are unique across regions; e.g. region "us-east-1" might be comprised of zones "us-east-1a" and "us-east-1b"

Reviewer 3: _optional tweak_
```suggestion
2) zone names are unique across regions; for example, a region "africa-east-1" might be comprised of zones "africa-east-1a" and "africa-east-1b"
```

Author: ACK :)

It should be safe to assume that topology labels do not change.  Even though labels are strictly mutable, consumers of them can assume that a given node is not going to be moved between zones without being destroyed and recreated.

> Kubernetes can use this information in various ways.  For example, the scheduler will automatically try to spread the Pods in a replication controller across nodes in a single-zone cluster (to reduce the impact of node failures, see [kubernetes.io/hostname](#kubernetesiohostname)). With multiple-zone clusters, this spreading behaviour also applies to zones (to reduce the impact of zone failures). This is achieved via _SelectorSpreadPriority_.

Reviewer 3: ```suggestion
Kubernetes can use this information in various ways.  For example, the scheduler automatically tries to spread the Pods belonging to a ReplicaSet across different nodes in a single-zone cluster (to reduce the impact of node failures, see [kubernetes.io/hostname](#kubernetesiohostname)). With multiple-zone clusters, this spreading behavior also applies to zones (to reduce the impact of zone failures). This is achieved via _SelectorSpreadPriority_.
```

Author: "in a ReplicaSet" ?

A region represents a larger domain, made up of one or more zones.  It is uncommon for Kubernetes clusters to span multiple regions,  While the exact definition of a zone or region is left to infrastructure implementations, common properties of a region include higher network latency between them than within them, non-zero cost for network traffic between them, and failure independence from other zones or regions.  For example, nodes within a region might share power infrastructure (e.g. a UPS or generator), but nodes in different regions typically would not.

Kubernetes makes a few assumptions about the structure of zones and regions:
1) regions and zones are hierarchical: zones are strict subsets of regions and no zone can be in 2 regions
2) zone names are unique across regions; for example region "africa-east-1" might be comprised of zones "africa-east-1a" and "africa-east-1b"

> It should be safe to assume that topology labels do not change.  Even though labels are strictly mutable, consumers of them can assume that a given node is not going to be moved between zones without being destroyed and recreated.

Reviewer 4: You might just include this last point in the numbered list of assumptions above. It is an important and well-written point.

Reviewer 3: Maybe a callout (eg `{{< note >}}`). It is an important point.
